// atlassian-connect-express/test/test_helper.js
process.env.AC_PRIVATE_KEY = "-----BEGIN RSA PRIVATE KEY-----\n\
MIIEpAIBAAKCAQEArVjfm73LwTC/S8M/Mwx9RYhNdEYqodULD/+jrT+Ser4PhcYw\n\
51DY3xbR3boZIn0GdNzVzWfoXoaiq24hvNsnxJc4pRyUn8RKr3DjTdqeuR64CWTq\n\
AckNjXUpWscj2ryCnLpl6zAbE3agE96wrt4KeBGxnvriZaROEpJmWpNt7f4NdKww\n\
uLFB7ci9L0xY1AbE7UHDwiJJ9B+n1Gk1xEeHPnzpPN4HFI7M9GYMWpzP+9BuHuiF\n\
v4aVBawbd+BxFQhQ+m5kELGmX9zW+Tcj/pOGY3kfGxZ6TekHE22TUM8v4wNbNbB+\n\
QrXV2dd2Q4w0JfPSQR8m6aSDO2EmjnLQSNcEZQIDAQABAoIBABxnqsSHdkLi7QXK\n\
KaoAyupA6og6EzRPh6+IcVarSyJn0H1YO5075z2vLJaLryCyogM2SKaO90m+ABvd\n\
QRQESWDycqrDoysG7EIRWn3gbA/jk76eVaYTPvBGGY+xgBOQvkpWVOsUB+vJjXOU\n\
Exmkcsy4AYYOkFidctmiOblv3RyZNqbDLrHtkgelq9tzi146ygaNhsD+EhRl/wW8\n\
zDFCYRFD8wgQcYOHO1r6PcsHFHTvT6fdymxOOGJJw13C+mhM/ZBaRKOB0iN5ieQ0\n\
DE/5dl0U2kgZWI4bhLhKmngls1uc+X9oAd/G5qpQ+5hLehnnu3pRt9vgbhrKwiJO\n\
p61GTuUCgYEA14OMiiCIwzEcONDAo2BlA4gH1gz2l5lzirdkDofkorYA33kg6paF\n\
ixjJJRTBO6sMOjb8HWPUroJ6zwcFDMJKcKU5nxD6mVrccA8HZD5NxfCRskPwtpTl\n\
FVU3cdxo8k0oGbNblFYvfbgBRCQ8RG/coRjUlyzB+12R9xxELQ0XHSsCgYEAzelz\n\
n1Ivsfv4yaoajW63qCR32m8XEyaJOmrC4mOuno1e+Nzmk4SS7S3XSfsmwQMVMP5S\n\
JI9ECXKGrAUlqwuA1FuZS8lH+Pz/5IZyxSltxK3Z4FaeuoKjbU0724ryR2Ad7dns\n\
8rOaqvcaWguAOTIgH00d6HMxlEHQf4qS5u+cPK8CgYEAyQS57QjFqBNBuLj2Ri2e\n\
avnSHVUAGY2XWRcP7AHTP/W1wLxR+J4nuAOJct32EzrLb+1joeJBOfkYpcxtaGY9\n\
Wu6yXW1P4gT/QkKyaU03694ld/qpJxSgzRIsYjbtaLyRH8/NJwJwz7n27GFEGa2h\n\
RF5NxSiJU3YGtVCXQtJH8fsCgYAYiDe+Q4xrKBn77Mg9EXPQFIgy/glBvtWaoB2X\n\
20hym2U1QRw3dxQKLev4ZHO5iyR7MdGXiXoRphNlazeRqU0v/7nDzbdFct8J2vrR\n\
gp4RUKUklS1UCzxtSUZYYbV7y3fFXrFcOQm0VLxO/g/IwLATCp0FlLPTdgPhLiFS\n\
Yjm7AQKBgQDTnobIbHTu1pshVwrnTiDVC0Akp4n7o4tMf3J/hcthH/vjjicuy7Dr\n\
VodIoKZ2seX58PKlDyRjwLURdZReqaBvriqbaZEjYi6Alc/VpwUZxOEfVugoSCt7\n\
pDUMvuFeY29lOKCQUDeszDrmFUMST8yHunacrWjx1yEm9AhIRYivaQ==\n\
-----END RSA PRIVATE KEY-----";

process.env.AC_PUBLIC_KEY = "-----BEGIN PUBLIC KEY-----\n\
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEArVjfm73LwTC/S8M/Mwx9\n\
RYhNdEYqodULD/+jrT+Ser4PhcYw51DY3xbR3boZIn0GdNzVzWfoXoaiq24hvNsn\n\
xJc4pRyUn8RKr3DjTdqeuR64CWTqAckNjXUpWscj2ryCnLpl6zAbE3agE96wrt4K\n\
eBGxnvriZaROEpJmWpNt7f4NdKwwuLFB7ci9L0xY1AbE7UHDwiJJ9B+n1Gk1xEeH\n\
PnzpPN4HFI7M9GYMWpzP+9BuHuiFv4aVBawbd+BxFQhQ+m5kELGmX9zW+Tcj/pOG\n\
Y3kfGxZ6TekHE22TUM8v4wNbNbB+QrXV2dd2Q4w0JfPSQR8m6aSDO2EmjnLQSNcE\n\
ZQIDAQAB\n\
-----END PUBLIC KEY-----";

exports.productPublicKey = 'MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEArVjfm73LwTC/S8M/Mwx9RYhNdEYqodULD/+jrT+Ser4PhcYw51DY3xbR3boZIn0GdNzVzWfoXoaiq24hvNsnxJc4pRyUn8RKr3DjTdqeuR64CWTqAckNjXUpWscj2ryCnLpl6zAbE3agE96wrt4KeBGxnvriZaROEpJmWpNt7f4NdKwwuLFB7ci9L0xY1AbE7UHDwiJJ9B+n1Gk1xEeHPnzpPN4HFI7M9GYMWpzP+9BuHuiFv4aVBawbd+BxFQhQ+m5kELGmX9zW+Tcj/pOGY3kfGxZ6TekHE22TUM8v4wNbNbB+QrXV2dd2Q4w0JfPSQR8m6aSDO2EmjnLQSNcEZQIDAQAB';
exports.productBaseUrl = "http://admin:admin@localhost:3001/confluence";

exports.addonPort = 3001;
exports.addonBaseUrl = "http://localhost:" + exports.addonPort;

exports.installedPayload = {
    "baseUrl": this.productBaseUrl,
    "key": "my add-on key",
    "clientKey": "clientKey",
    "sharedSecret": "sharedSecret",
    "publicKey": this.productPublicKey,
    "eventType": "installed"
};

exports.clientInfo = {
  baseUrl: this.addonBaseUrl,
  publicKey: this.productPublicKey,
  description: 'host.consumer.default.description',
  pluginsVersion: '0.6.1010',
  clientKey: 'testHostClientKey',
  serverVersion: '4307',
  key: 'testKey',
  productType: 'test'
};

exports.consumerInfo = '\
<?xml version="1.0" encoding="UTF-8"?>\n\
<consumer>\n\
  <key>testHostClientKey</key>\n\
  <name>Confluence</name>\n\
  <publicKey>' + this.productPublicKey + '</publicKey>\n\
  <description>Atlassian Confluence</description>\n\
</consumer>\
';

// Allows us to run tests from a different dir
process.chdir(__dirname);
